src/main/kotlin/org/jqassistant/tooling/intellij/plugin/editor/rules/refactoring/delete/RuleSafeDeleteProcessorDelegate.kt

```diff
@@ -0,0 +1,256 @@
+     *
+     * @param element an element selected for deletion.
+     * @param allElementsToDelete all elements selected for deletion.
+     * @return additional elements to search for usages, or null if the user has cancelled the refactoring.
+     */
+    override fun getElementsToSearch(
+        element: PsiElement,
+        allElementsToDelete: MutableCollection<out PsiElement>,
+    ): MutableCollection<out PsiElement> {
+        // Adding elements makes the refactoring not even delete the element it was called on
+        val result: MutableCollection<PsiElement> = mutableListOf(element)
+        return result
+    }
+
+    /**
+     * Returns the list of additional elements to be deleted. Called after the refactoring dialog is shown.
+     * May show UI to ask the user if some additional elements should be deleted along with the
+     * specified selected element.
+     * This implementation returns all `providesConcept` tags that reference the selected element.
+     *
+     * @param element an element selected for deletion.
+     * @param allElementsToDelete all elements selected for deletion.
+     * @return additional elements to search for usages, or null if no additional elements were chosen.
+     */
+    override fun getAdditionalElementsToDelete(
+        element: PsiElement,
+        allElementsToDelete: MutableCollection<out PsiElement>,
+        askUser: Boolean,
+    ): MutableCollection<PsiElement> {
+        val files = findEligiblePsiFiles(element)
+        val result = mutableSetOf<PsiElement>()
+        for (file in files) {
+            result.addAll(findRefIdUsages(element, file, "providesConcept"))
+            result.addAll(findRefIdUsages(element, file, "requiresConcept"))
+            result.addAll(findRefIdUsages(element, file, "includeConcept"))
+            result.addAll(findRefIdUsages(element, file, "includeConstraint"))
+            result.addAll(findRefIdUsages(element, file, "includeGroup"))
+        }
+        return result
+    }
+
+    /**
+     * Detects usages which are not safe to delete.
+     * TODO when looking at the usages after dialog is shown by IntelliJ, IDE (2023.3) throws an error unrelated to this code
+     *
+     * @param element an element selected for deletion.
+     * @param allElementsToDelete all elements selected for deletion.
+     * @return collection of conflict messages which would be shown to the user before delete can be performed.
+     */
+    override fun findConflicts(
+        element: PsiElement,
+        allElementsToDelete: Array<out PsiElement>,
+    ): MutableCollection<String> {
+        /*
+         * Explicitly check for conflicts in the context of the tag, its parent and the $element to be deleted
+         * */
+        val messages = mutableListOf<String>()
+        val possibleConflicts =
+            allElementsToDelete.filter { (it as? XmlTag?)?.name == "requiresConcept" }.map { it as XmlTag }
```

Comment:
```suggestion
            allElementsToDelete.filterIsInstance<XmlTag>().filter { it.name == "requiresConcept" }
```
Nit: writing it this way wsould be more idiomatic, you can also take a look at [mapNotNull](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/map-not-null.html), [filterNotNull](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/filter-not-null.html) and [filterIsInstance](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/filter-is-instance.html)